engine/src/main/java/org/kigalisim/engine/SingleThreadEngine.java

```diff
@@ -456,39 +762,342 @@ private UnitConverter createUnitConverterWithTotal(String stream) {
+        percentImport = importInitialChargeVal.divide(totalInitialChargeVal, 10, java.math.RoundingMode.HALF_UP);
+      }
+    } else {
+      percentManufacture = totalNonRecycleKg.compareTo(BigDecimal.ZERO) == 0 
+          ? BigDecimal.ZERO : manufactureSalesKg.divide(totalNonRecycleKg, 10, java.math.RoundingMode.HALF_UP);
+      percentImport = totalNonRecycleKg.compareTo(BigDecimal.ZERO) == 0 
+          ? BigDecimal.ZERO : importSalesKg.divide(totalNonRecycleKg, 10, java.math.RoundingMode.HALF_UP);
+    }
+
+    // Recycle
+    EngineNumber newRecycleValue = new EngineNumber(recycledDisplacedKg, "kg");
+    this.streamKeeper.setStream(application, substance, "recycle", newRecycleValue);
+
+    // New values
+    BigDecimal requiredKgUnbound = kgForRecharge.add(kgForNew);
+    BigDecimal requiredKg = requiredKgUnbound.compareTo(BigDecimal.ZERO) < 0 ? BigDecimal.ZERO : requiredKgUnbound;
+    BigDecimal newManufactureKg = percentManufacture.multiply(requiredKg);
+    BigDecimal newImportKg = percentImport.multiply(requiredKg);
+    EngineNumber newManufacture = new EngineNumber(newManufactureKg, "kg");
+    EngineNumber newImport = new EngineNumber(newImportKg, "kg");
+    this.streamKeeper.setStream(application, substance, "manufacture", newManufacture);
+    this.streamKeeper.setStream(application, substance, "import", newImport);
   }
 
   /**
-   * Placeholder for retirement recalculation.
+   * Recalculates retirement values based on current state.
    *
    * @param scope The scope to recalculate for
    */
   private void recalcRetire(Scope scope) {
-    // TODO: Implement recalc logic
+    // Setup
+    OverridingConverterStateGetter stateGetter =
+        new OverridingConverterStateGetter(this.stateGetter);
+    UnitConverter unitConverter = new UnitConverter(stateGetter);
+    Scope scopeEffective = scope != null ? scope : this.scope;
+    String application = scopeEffective.getApplication();
+    String substance = scopeEffective.getSubstance();
+
+    // Check allowed
+    if (application == null || substance == null) {
+      throw new RuntimeException(
+          "Tried recalculating population change without application and substance.");
+    }
+
+    // Calculate change
+    EngineNumber currentPriorRaw = this.streamKeeper.getStream(application, substance, "priorEquipment");
+    EngineNumber currentPrior = unitConverter.convert(currentPriorRaw, "units");
+
+    EngineNumber currentEquipmentRaw = this.streamKeeper.getStream(application, substance, "equipment");
+    EngineNumber currentEquipment = unitConverter.convert(currentEquipmentRaw, "units");
+
+    stateGetter.setPopulation(currentPrior);
+    EngineNumber amountRaw = this.streamKeeper.getRetirementRate(application, substance);
+    EngineNumber amount = unitConverter.convert(amountRaw, "units");
+    stateGetter.clearPopulation();
+
+    EngineNumber newPrior = new EngineNumber(
```

Comment:
Please simplify with:

```
BigDecimal newPriorValue = currentPrior.getValue().subtract(amount.getValue());
```

Please apply the same for `newEquipment` using `newEquipmentValue`